docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,301 @@
+for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
+limits and requests are handled for pods).
+
+It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
+Each infrastructure provider will be responsible for determining the best implementation for adding the
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU ResourceName = "cpu"
+    AutoscalerResourceMemory ResourceName = "memory"
+    AutoscalerResourceGPU ResourceName = "gpu"
+)
+
+// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineStatus struct {
+    Capacity ResourceList `json:"capacity,omitempty"`
+}
+
+// AutoscalerMachineTemplateStatus defines information that is used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineTemplateStatus struct {
+    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
+}
+
+// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
+type DockerMachineTemplateStatus struct {
+    Autoscaler AutoscalerMachineTemplateStatus `json:"autoscaler,omitempty"`
+}
+
+// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
+type DockerMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
+    Status DockerMachineTemplateStatus `json:"status,omitempty"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  template:
+    spec: {}
+status:
+  autoscaler:
+    machine:
+      capacity:
+        memory: 500mb
+        cpu: "1"
+        gpu: "1"
+```
+
+#### MachineSet and MachineDeployment Annotations
+
+In cases where a user needs to provide specific resource information for a
+MachineSet or MachineDeployment, or in cases where an infrastructure provider
+has not yet added the Infrastructure Machine Template status changes, they
+may use annotations to provide the information.
+
+If a user wishes to specify the resource capacity through annotations, they
+may do by adding the following to any MachineSet or MachineDeployment (it is not required on both)
+that are participating in autoscaling:
+
+```
+kind: <MachineSet or MachineDeployment>
+metadata:
+  annotations:
+      autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "1"
+      autoscaler.cluster.x-k8s.io/machine-capacity-memoryMb: "500"
```

Comment:
```suggestion
      autoscaler.cluster.x-k8s.io/machine-capacity-memory: "500mb"
```
nit: I wonder if we should have the unit in the value and not in the annotation name (would be consistent with the field / the pod request/limit fields)

Reply:
(last nit from my side, I promise :))

Reply:
my feeling here is that if we drop the `Mb` from the name then the value should be whatever a `resource.Quantity` could handle. adding it to the annotation key makes it explicit what is expected here.

Reply:
Yup, just think its inconsistent with the corresponding field. But feel free to keep it as is.

Reply:
i don't have a strong feeling about keeping the `Mb` in the annotation, but i want to confer with a colleague to see if they had another reason for suggesting that name. if not, i'm happy to change it and just say we will allow `resource.Quantity` in that field (i suppose the cpu field would follow the same pattern then).

Reply:
Oh, good point that cpu could also be a quantity. I'm not sure if cpu has to be an integer though. Are there cases where we don't have "whole" cpus?

Reply:
ok, had some conversations about these annotations and have update the memory one to be more consistent with the structured API.